File: recap-subworker/recap_subworker/app/infra/pki/metrics.py
"""Lifecycle metrics for in-process enrollment on a private registry."""

from __future__ import annotations

import time

from prometheus_client import REGISTRY as DEFAULT_REGISTRY
from prometheus_client import CollectorRegistry, Counter, Gauge

from recap_subworker.app.infra.pki.state import State


class PromObserver:
    """Publishes pki_enrollment_* matching the Go cohort metric family."""

    def __init__(self, subject: str, registry: CollectorRegistry) -> None:
        if registry is DEFAULT_REGISTRY:
            raise TypeError("pki: PromObserver must use a private CollectorRegistry")
        self._subject = subject
        self._remaining = Gauge(
            "cert_remaining_seconds",
            "Seconds until the current leaf expires. Negative if expired.",
            labelnames=("subject",),
            namespace="pki_enrollment",
            registry=registry,
        ).labels(subject=subject)
        self._not_after = Gauge(
            "cert_not_after_seconds",
            "Unix timestamp of the current leaf certificate's not_after.",
            labelnames=("subject",),
            namespace="pki_enrollment",
            registry=registry,
        ).labels(subject=subject)
        self._last_rotation = Gauge(
            "last_rotation_timestamp_seconds",
            "Unix timestamp of the last successful cert rotation.",
            labelnames=("subject",),
            namespace="pki_enrollment",
            registry=registry,
        ).labels(subject=subject)
        self._healthy = Gauge(
            "healthy",
            "1 if the cert on disk is currently valid (not expired).",
            labelnames=("subject",),
            namespace="pki_enrollment",
            registry=registry,
        ).labels(subject=subject)
        self._renewal_total = Counter(
            "renewal_total",
            "Count of completed rotation attempts grouped by outcome.",
            labelnames=("result", "subject"),
            namespace="pki_enrollment",
            registry=registry,
        )
        self._reissue_total = Counter(
            "reissue_total",
            "Count of reissuances by reason (missing / expired / near_expiry / corrupt).",
            labelnames=("reason", "subject"),
            namespace="pki_enrollment",
            registry=registry,
        )
        self._loop_error_total = Counter(
            "loop_error_total",
            "Unexpected renewal-loop exceptions that were logged and retried.",
            labelnames=("subject",),
            namespace="pki_enrollment",
            registry=registry,
        )
        self._state = State.MISSING

    def on_classified(self, state: State, remaining: float) -> None:
        self._state = state
        self._remaining.set(remaining)
        self._not_after.set(time.time() + remaining)
        if state in {State.EXPIRED, State.CORRUPT, State.MISSING}:
            self._healthy.set(0)
            return
        self._healthy.set(1)

    def on_reissued(self, reason: str) -> None:
        self._reissue_total.labels(reason=reason, subject=self._subject).inc()

    def on_renewed(self, success: bool) -> None:
        result = "success" if success else "failure"
        if success:
            self._last_rotation.set(time.time())
        self._renewal_total.labels(result=result, subject=self._subject).inc()

    def on_retry(self, attempt: int, err: BaseException) -> None:
        del attempt, err
        self._loop_error_total.labels(subject=self._subject).inc()

File: rag-orchestrator/internal/pki/metrics.go
package pki

import (
	"sync"
	"time"

	"github.com/prometheus/client_golang/prometheus"
	"github.com/prometheus/client_golang/prometheus/promauto"
)

// PromObserver implements Observer and publishes the in-process enrollment
// metric family on a caller-supplied registry. rag-orchestrator scrapes PKI
// series from the dedicated ops listener (:9110), never from Echo /metrics.
type PromObserver struct {
	subject string

	notAfter     prometheus.Gauge
	remaining    prometheus.Gauge
	lastRotation prometheus.Gauge
	renewalTotal *prometheus.CounterVec
	reissueTotal *prometheus.CounterVec
	healthy      prometheus.Gauge

	mu    sync.RWMutex
	state State
}

// NewPromObserver registers collectors on reg (DefaultRegisterer when nil).
// Tests pass a private registry so they never collide with production.
func NewPromObserver(subject string, reg prometheus.Registerer) *PromObserver {
	if reg == nil {
		reg = prometheus.DefaultRegisterer
	}
	factory := promauto.With(reg)
	labels := prometheus.Labels{"subject": subject}
	return &PromObserver{
		subject: subject,
		notAfter: factory.NewGauge(prometheus.GaugeOpts{
			Namespace: "pki_enrollment", Name: "cert_not_after_seconds",
			Help:        "Unix timestamp of the current leaf certificate's not_after.",
			ConstLabels: labels,
		}),
		remaining: factory.NewGauge(prometheus.GaugeOpts{
			Namespace: "pki_enrollment", Name: "cert_remaining_seconds",
			Help:        "Seconds until the current leaf expires. Negative if expired.",
			ConstLabels: labels,
		}),
		lastRotation: factory.NewGauge(prometheus.GaugeOpts{
			Namespace: "pki_enrollment", Name: "last_rotation_timestamp_seconds",
			Help:        "Unix timestamp of the last successful cert rotation.",
			ConstLabels: labels,
		}),
		renewalTotal: factory.NewCounterVec(prometheus.CounterOpts{
			Namespace: "pki_enrollment", Name: "renewal_total",
			Help:        "Count of completed rotation attempts grouped by outcome.",
			ConstLabels: labels,
		}, []string{"result"}),
		reissueTotal: factory.NewCounterVec(prometheus.CounterOpts{
			Namespace: "pki_enrollment", Name: "reissue_total",
			Help:        "Count of reissuances by reason (missing / expired / near_expiry / corrupt).",
			ConstLabels: labels,
		}, []string{"reason"}),
		healthy: factory.NewGauge(prometheus.GaugeOpts{
			Namespace: "pki_enrollment", Name: "healthy",
			Help:        "1 if the cert on disk is currently valid (not expired).",
			ConstLabels: labels,
		}),
	}
}

func (o *PromObserver) OnClassified(state State, remaining time.Duration) {
	o.mu.Lock()
	o.state = state
	o.mu.Unlock()
	o.remaining.Set(remaining.Seconds())
	o.notAfter.Set(float64(time.Now().Add(remaining).Unix()))
	if state == StateExpired || state == StateCorrupt || state == StateMissing {
		o.healthy.Set(0)
		return
	}
	o.healthy.Set(1)
}

func (o *PromObserver) OnReissued(reason string) {
	o.reissueTotal.WithLabelValues(reason).Inc()
}

func (o *PromObserver) OnRenewed(success bool) {
	result := "success"
	if !success {
		result = "failure"
	} else {
		o.lastRotation.Set(float64(time.Now().Unix()))
	}
	o.renewalTotal.WithLabelValues(result).Inc()
}

func (o *PromObserver) OnRetry(int, error) {}
